File: bindings/deno/src/safe_header.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeHeader - HTTP header operations that prevent CRLF injection.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 *
 * @module
 */

import { getLib, ProvenStatus, readAndFreeString, statusToError } from './ffi.ts';
import { err, ok, type Result } from './result.ts';

/**
 * Safe HTTP header operations backed by formally verified Idris 2 code.
 */
export class SafeHeader {
  /**
   * Check for CRLF injection characters in a header value.
   *
   * @param value - The header value to check.
   * @returns Result containing a boolean or an error string.
   */
  static hasCrlf(value: string): Result<boolean> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(value);
    const result = symbols.proven_header_has_crlf(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }

  /**
   * Check if a header name is a valid token per RFC 7230.
   *
   * @param name - The header name to validate.
   * @returns Result containing a boolean or an error string.
   */
  static isValidName(name: string): Result<boolean> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(name);
    const result = symbols.proven_header_is_valid_name(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }

  /**
   * Check if a header name is in the dangerous headers list.
   *
   * @param name - The header name to check.
   * @returns Result containing a boolean or an error string.
   */
  static isDangerous(name: string): Result<boolean> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(name);
    const result = symbols.proven_header_is_dangerous(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }

  /**
   * Create a validated header string "Name: Value".
   *
   * @param name - The header name.
   * @param value - The header value.
   * @returns Result containing the rendered header or an error string.
   */
  static render(name: string, value: string): Result<string> {
    const symbols = getLib();
    const nameBytes = new TextEncoder().encode(name);
    const valueBytes = new TextEncoder().encode(value);
    const result = symbols.proven_header_render(
      nameBytes,
      nameBytes.length,
      valueBytes,
      valueBytes.length,
    );
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    const rendered = readAndFreeString(result[1], Number(result[2]));
    if (rendered === null) return err('Null string returned');
    return ok(rendered);
  }

  /**
   * Build Content-Security-Policy header value from directives JSON.
   *
   * @param directivesJson - JSON string of CSP directives.
   * @returns Result containing the CSP header value or an error string.
   */
  static buildCsp(directivesJson: string): Result<string> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(directivesJson);
    const result = symbols.proven_header_build_csp(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    const csp = readAndFreeString(result[1], Number(result[2]));
    if (csp === null) return err('Null string returned');
    return ok(csp);
  }

  /**
   * Build HSTS (Strict-Transport-Security) header value.
   *
   * @param maxAge - Max age in seconds.
   * @param includeSubdomains - Whether to include subdomains.
   * @param preload - Whether to include preload directive.
   * @returns Result containing the HSTS header value or an error string.
   */
  static buildHsts(
    maxAge: number,
    includeSubdomains: boolean,
    preload: boolean,
  ): Result<string> {
    const symbols = getLib();
    const result = symbols.proven_header_build_hsts(
      BigInt(maxAge),
      includeSubdomains,
      preload,
    );
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    const hsts = readAndFreeString(result[1], Number(result[2]));
    if (hsts === null) return err('Null string returned');
    return ok(hsts);
  }
}

File: bindings/swift/Sources/Proven/SafeHeader.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Safe HTTP header operations delegated to libproven FFI.
///
/// All validation and rendering is performed by the formally verified
/// Idris 2 core via the Zig FFI bridge.

import CProven

public enum SafeHeader {
    /// Check if a header value contains CRLF injection characters.
    public static func hasCRLF(_ value: String) -> Result<Bool, ProvenError> {
        withStringBytes(value) { ptr, len in
            let result = proven_header_has_crlf(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(result.value)
        }
    }

    /// Check if a header name is valid per RFC 7230.
    public static func isValidName(_ name: String) -> Result<Bool, ProvenError> {
        withStringBytes(name) { ptr, len in
            let result = proven_header_is_valid_name(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(result.value)
        }
    }

    /// Check if a header name is considered dangerous (e.g. security-sensitive).
    public static func isDangerous(_ name: String) -> Result<Bool, ProvenError> {
        withStringBytes(name) { ptr, len in
            let result = proven_header_is_dangerous(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(result.value)
        }
    }

    /// Render a validated header as "Name: Value" string.
    public static func render(name: String, value: String) -> Result<String, ProvenError> {
        let nameUtf8 = Array(name.utf8)
        let valueUtf8 = Array(value.utf8)
        return nameUtf8.withUnsafeBufferPointer { nameBuf in
            valueUtf8.withUnsafeBufferPointer { valueBuf in
                let result = proven_header_render(
                    nameBuf.baseAddress, nameBuf.count,
                    valueBuf.baseAddress, valueBuf.count
                )
                return consumeStringResult(result)
            }
        }
    }

    /// Build a Content-Security-Policy header from JSON-encoded directives.
    public static func buildCSP(directivesJSON: String) -> Result<String, ProvenError> {
        withStringBytes(directivesJSON) { ptr, len in
            consumeStringResult(proven_header_build_csp(ptr, len))
        }
    }

    /// Build a Strict-Transport-Security header.
    public static func buildHSTS(
        maxAge: Int64,
        includeSubdomains: Bool = false,
        preload: Bool = false
    ) -> Result<String, ProvenError> {
        let result = proven_header_build_hsts(maxAge, includeSubdomains, preload)
        return consumeStringResult(result)
    }
}
